core/Enclave/key_factory.cpp

```diff
@@ -0,0 +1,301 @@
+}
+
+static uint32_t ehsm_get_public_key_pem_size(const uint32_t keyspec)
+{
+    switch(keyspec)
+    {
+        case EH_RSA_2048:
+            return RSA_2048_PUBLIC_KEY_PEM_SIZE;
+        case EH_RSA_3072:
+            return RSA_3072_PUBLIC_KEY_PEM_SIZE;
+        case EH_RSA_4096:
+            return RSA_4096_PUBLIC_KEY_PEM_SIZE;
+        case EH_EC_P256:
+        case EH_SM2:
+            return ECC_PUBLIC_KEY_PEM_SIZE;
+        default:
+            return 0;
+    }
+}
+
+sgx_status_t ehsm_create_keyblob(const uint32_t plaintext_size, const uint8_t *plaintext,
+                                        const uint32_t aad_size, const uint8_t *aad,
+                                        const uint32_t gcm_data_size, sgx_aes_gcm_data_ex_t *gcm_data)
+{
+    uint32_t real_aad_size = aad_size;
+    if (NULL == aad)
+        real_aad_size = 0;
+
+    sgx_status_t ret = sgx_read_rand(gcm_data->iv, sizeof(gcm_data->iv));
+    if (ret != SGX_SUCCESS) {
+        printf("error generating iv.\n");
+        return ret;
+    }
+
+    ret = sgx_rijndael128GCM_encrypt(&g_domain_key,
+                                     plaintext, plaintext_size,
+                                     gcm_data->payload,
+                                     gcm_data->iv, sizeof(gcm_data->iv),
+                                     &(gcm_data->payload[gcm_data->ciphertext_size]), real_aad_size,
+                                     reinterpret_cast<uint8_t (*)[16]>(gcm_data->mac));
+    if (SGX_SUCCESS != ret) {
+        printf("gcm encrypting failed.\n");
+    }
+    else {
+        gcm_data->ciphertext_size = plaintext_size;
+        gcm_data->aad_size = real_aad_size;
+    }
+
+    return ret;
+}
+
+sgx_status_t ehsm_parse_keyblob(uint32_t plaintext_size, uint8_t *plaintext,
+                                        const sgx_aes_gcm_data_ex_t *gcm_data)
+{
+    if (NULL == gcm_data || NULL == plaintext || NULL == plaintext_size
+                 || plaintext_size < gcm_data->ciphertext_size)
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    sgx_status_t ret = sgx_rijndael128GCM_decrypt(&g_domain_key,
+                                                  gcm_data->payload, gcm_data->ciphertext_size,
+                                                  plaintext,
+                                                  gcm_data->iv, sizeof(gcm_data->iv),
+                                                  &(gcm_data->payload[gcm_data->ciphertext_size]), gcm_data->aad_size,
+                                                  (const sgx_aes_gcm_128bit_tag_t*)gcm_data->mac);
+    if (SGX_SUCCESS != ret)
+        printf("gcm decrypting failed.\n");
+    else
+        plaintext_size = gcm_data->ciphertext_size;
+
+    return ret;
+}
+
+sgx_status_t ehsm_create_aes_key(ehsm_keyblob_t *cmk)
+{
+    return SGX_ERROR_UNEXPECTED;
+}
+
+sgx_status_t ehsm_create_asymmetric_key(ehsm_keyblob_t *cmk)
+{
+    sgx_status_t ret = SGX_SUCCESS;
+
+    EVP_PKEY_CTX        *keypair        = NULL;
```

Comment:
why call it as keypair? seems that it's a context right?

so rename it as ctx?

Reply:
it fixed